x-pack/plugins/actions/server/saved_objects/schemas/raw_connector/v1.ts

```diff
@@ -11,8 +11,9 @@ export const rawConnectorSchema = schema.object({
   actionTypeId: schema.string(),
   name: schema.string(),
   isMissingSecrets: schema.boolean(),
-  config: schema.recordOf(schema.string(), schema.any()),
-  secrets: schema.recordOf(schema.string(), schema.any()),
+  config: schema.any({ defaultValue: {} }),
+  secrets: schema.any({ defaultValue: {} }),
+
```

Comment:
We have used empty objects (`config: {}`) for config and secrets in functional tests (probably in some old data as well)  but `schema.recordOf` cannot validate them and crushes the kibana. I tried to use `schema.oneOf` as well but it didn't work either. `schema.any` is the only definition that works for both empty and populated objects.

Reply:
For me the schema with empty objects works and `recordOf`
```
const { config, secrets, ...rest } = action;
rawConnectorSchema.validate({ ...rest, config: {}, secrets: {} })
```

probably it was some specific test?

Reply:
My comment here somehow disappeared, adding again:

Thanks for reviewing @elena-shostak.

hundreds of functional tests are failing when I use `schema.recordOf`
Below is one of them, maybe you can test with it?

https://github.com/elastic/kibana/blob/3a217fd7c79391bcc02cf46f5ebcaa8369042505/x-pack/test/alerting_api_integration/security_and_spaces/group1/tests/alerting/find.ts#L37

Reply:
@elena-shostak It looks like what's happening is that because `secrets` is encrypted, it's saved in the saved object as an encrypted string, so the schema is validating that encrypted string as an object and running into the validation error. What is the right way to do this for encrypted saved objects?

Reply:
Encrypted attributes will always be a string, but decrypted attributes could have some specific schema.

> hundreds of functional tests are failing

Can you point to where the failing test are? If they operate on encrypted objects, the encrypted attributes will always be strings.

As for the model version schema, objects are decrypted prior to migration (see https://github.com/elastic/kibana/pull/174238).

So it may be that what you need is one schema for the encrypted object (where your tests are failing) and a separate one for the decrypted object (to use in the model version, or anywhere you want to verify the decrypted object). Someone from @elastic/kibana-core might have better advice to offer.

Reply:
Encrypted attributes will always be a string, but decrypted attributes could have some specific schema.

> hundreds of functional tests are failing

Can you point to where the failing test are? If they operate on encrypted objects, the encrypted attributes will always be strings.

As for the model version schema, objects are decrypted prior to downward migration (see https://github.com/elastic/kibana/pull/174238), but not upward migration. So I am not sure what this means for applying the `create` schema. Maybe someone from @elastic/kibana-core can chime in.

It may be that what you need is one schema for the encrypted object (where your tests are failing) and a separate one for the decrypted object (to use in the model version `create` schema, or anywhere you want to verify the decrypted object). Someone from @elastic/kibana-core might have better advice to offer.

Reply:
Encrypted attributes will always be a string, but decrypted attributes could have some specific schema.

> hundreds of functional tests are failing

Can you point to where the failing test are? If they operate on encrypted objects, the encrypted attributes will always be strings.

As for the model version schema, objects are decrypted prior to downward migration (see https://github.com/elastic/kibana/pull/174238), but not upward migration. So I am not sure what this means for applying the `create` schema, though I assume this is used only when creating a new object from scratch, and not applied to objects that are migrated via the model version changes. Maybe someone from @elastic/kibana-core can chime in.

It may be that what you need is one schema for the encrypted object (where your tests are failing) and a separate one for the decrypted object (to use in the model version `create` schema, or anywhere you want to verify the decrypted object). Someone from @elastic/kibana-core might have better advice to offer.

Reply:
Hi @jeramysoucy 

All the tests that we use below requests are failing when i use `schema.recordOf`. 

```
const { body: createdAction } = await supertest
        .post(`${getUrlPrefix(Spaces.space1.id)}/api/actions/connector`)
        .set('kbn-xsrf', 'foo')
        .send({
          name: 'MY action',
          connector_type_id: 'test.noop',
          config: {},
          secrets: {},
        })
        .expect(200);
```

And when i debug it i get `Error: [attributes.secrets]: could not parse record value from json input: Bad Request` error as a result of validation from modelVersions.create schema.

Reply:
> So I am not sure what this means for applying the create schema, though I assume this is used only when creating a new object from scratch, and not applied to objects that are migrated via the model version changes

@jeramysoucy you are correct, model versions are only created when one doesn't already exist. All subsequent migrations are changes

Reply:
@ersin-erdal @ymao1 
I took a deeper look. There's a lot of abstraction happening in the actions code. Here are some observations...
- From what I can tell, the original `raw_connector_schema` was never referenced anywhere in the code except for a unit test for the schema itself. Can you confirm that this is true? If so, we're not doing ourselves any favors in comparing the old definition to the new one. The new versioned schema is actually being used in the new model version, so it makes sense that the definition may need change to accommodate this.
- Action secrets seem to go through a whole schema validation process in the `ActionsClient`, where the schema for a specific type of action is validated. This is before we ever get to the CRUD operations at the SO level, where we need to be able to accommodate any shape the secrets attribute might take.
- I need @elastic/kibana-core to confirm once more, but the model version create schema appears to need to account for the encrypted shape of the object, not the decrypted one. Is this true? If so, then the schema should be `schema.string()`, and if not it needs to be something that accounts for the secrets schema of any `action` type.
- Side note: I am a little confused by the nomenclature. What is the relationship between a "raw connector" and an "action"? I see the SO type is called an `action`, but the schema used for the action's model version is called `rawConnectorSchema`? Is this just a matter of legacy?

Reply:
@jeramysoucy 

> From what I can tell, the original raw_connector_schema was never referenced anywhere in the code except for a unit test for the schema itself. Can you confirm that this is true? 

It was added when we added some validation on task run but we have since reverted those validation changes so you are correct that currently that schema is not being used

> Action secrets seem to go through a whole schema validation process in the ActionsClient,

This is correct

> Side note: I am a little confused by the nomenclature.

Sorry for the confusion! The `actions` SO type is a legacy name for connectors. We've been trying to use the correct naming convention in the code but cannot change the SO type naming.

Reply:
So it sounds like we should be using `schema.string()` for the secrets in the connector schema because the encrypted SO saves secrets as a string.

Reply:
Sorry for the confusion, as Ying said this is the first time we use `raw_connector_schema` for model versions.
Therefore the existing tests were failing.
Looks like using `schema.string()` for the secrets field solved our problem, thanks for helping us here.